# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
def addTwo(x):
    return x + 2

print(addTwo(4))
print((lambda x: x+2)(4)) #pass in data
a_num = 4 #var
print((lambda x: x +2)(a_num)) 

6
6
6


In [10]:
print((lambda i: i+ 2 * i)(4))

12


#### Saving to a Variable

In [2]:
f_test = lambda x: x + 2
f_test(4)

6

In [8]:
l_test = lambda i: i * 2
l_test(4)

8

#### Multiple Inputs

In [14]:
print((lambda x, y, z: x * y * z)(3, 5, 8))
x_test = lambda x, y, z: x * y * z
print(x_test(3, 5, 8))

120
120


#### Passing a Lambda into a Function

In [12]:
def multiply(f, num):
    """
        f expects a lambda function
        num expects a number
    
    """
    return f(num)
multiply(lambda x: x * x, 4)

16

#### Returning a Lambda from a Function

In [13]:
#regular
def multiply_test(num):
    return num * 4

#func within a func
def returnFunc():
    test = 4
    def multiply(num):
        return num * 2
    return multiply

f_return = returnFunc()
print(returnFunc())
print(f_return(4))


# returned from Regular Func because why do all that ^

def returnLam(b,c):
    return lambda x, a: x + a + b + c
r_lamb = returnLam(4,5)

print(r_lamb(5,5))

<function returnFunc.<locals>.multiply at 0x7fb3996b5cf0>
8
19


#### If Statements within Lambdas

In [20]:
#Lambda x: True if (condition) else False

fcondition = lambda num:num *2 if num > 10 else num + 2
print((fcondition(8)),"; 8 is less than 10")
print((fcondition(12)),"; 12 is greater than 10")
print((fcondition(10)),"; 10 is greater than 10 too")

10 ; 8 is less than 10
24 ; 12 is greater than 10
12 ; 10 is greater than 10 too


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [7]:
f = lambda cube: cube ** 3

print(f(2))
print(f(3))
print(f(4))

8
27
64


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. <br><b>To turn it into a list we use the list() type conversion.</b></p>

#### Syntax

> <i>map(func, iterable)</i>
    
   an iterable has to be something like a list, dict, tuple, set
   
   Normally the usage of map happens with a pre-defined function - but we can use lambas as well

In [25]:
def squared(num, num2):
    if num<10 and num <10:
        return num **2, num**2
    else:
        return num, num2
        
numbers = [4,11,20,3,15,20]
more_nums = [4,10, 3, 2, 6]

squared_nums_map= map(squared, numbers, more_nums)
#passing SQUAED as the function, the others are iterables
print(squared_nums_map) #map object
print(' ')
squared_nums_map= list(map(squared, numbers, more_nums))
print(squared_nums_map)

 
[(16, 16), (11, 10), (20, 3), (9, 9), (15, 6)]


#### Using Lambda's with Map

In [30]:
#map(lambda x: do something, iterable)
#using lambda in map happens inline or on one line

#list(map(lmabda x: x+2, nums))
squared_nums_lambda = list(map(lambda x, y: (x**2, y**2) if x<10 and y<10 else (x,y), numbers, more_nums))
print(squared_nums_lambda)

print(list(map(lambda x, y: (x**2, y**2) if x<10 and y<10 else (x,y), numbers, more_nums)))

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]
[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [33]:
#output should be 
#7,21,39,5,29,39
#7,19,5,3,11
#map(func,iterable)

print(list(map(lambda a, b: ((a*2)-1, (b*2)-1), numbers, more_nums)))
print(' ')
print(list(map(lambda a: (a * 2)-1, numbers)))
print(list(map(lambda b: (b * 2)-1, more_nums)))

[(7, 7), (21, 19), (39, 5), (5, 3), (29, 11)]
 
[7, 21, 39, 5, 29, 39]
[7, 19, 5, 3, 11]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed.
    <br> Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [36]:
names = ['Bob', 'Andy', 'Max', 'Evan', 'Angelica']

def a_names(name):
    if name[0].lower() == 'a':
        return True
    elif name[0].lower()== 'b':
        return True
    else: 
        return False
new_names_list = list(filter(a_names, names))
print(new_names_list)
print(a_names)

['Bob', 'Andy', 'Angelica']
<function a_names at 0x7fb39a395bd0>


#### Using Lambda's with Filter()

In [37]:
new_names_l= list(filter(lambda name: True if name[0].lower() == 'a' else False, names))
print(new_names_l)

['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [52]:
import statistics as st
more_nums = [4,10, 3, 2, 6]

finding_mean = st.mean([4,10, 3, 2, 6])
print(finding_mean)

#find_nums_below = list(filter(lambda nums: True if nums < finding_mean else False, finding_mean))

find_nums_below = list(filter(lambda num: num < finding_mean, more_nums))
print(find_nums_below)
print(list(filter(lambda num: num < finding_mean, more_nums)))

5
[4, 3, 2]
[4, 3, 2]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [54]:
from functools import reduce 

# reduce(function, iterable)

list_1 = [2, 4, 6, 8, 10]

def addNums(num1, num2):
    return num1 + num2

result_add = reduce(addNums, list_1)
print(result_add)

# subtract a list of numbers
def subtractNums(num1, num2):
    return num1 - num2

result_sub = reduce(subtractNums, list_1)
print(result_sub)

30
-26


#### Using Lambda's with Reduce()

In [55]:
result_l = reduce(lambda x,y: x+y, list_1)
print(result_l)

30


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [56]:
my_list = [1,2,3,4]
mult_l = reduce(lambda x,y: x*y, my_list)
print(mult_l)

24


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [61]:
def addNums(num):
    #base case
    if num <= 1:
        print('addNums(1) = 1')
        return num
    else:
        print(f'addNums({num}) = {num} + addNums({num -1})')
        return num + addNums(num -1)
    
# def addNums(num):
#     print(f'addNums({num}) = {num} + addNums({num -1})')
#     return num + addNums(num -1)
# #RecursionError: maximum recursion depth exceeded while calling a Python object
    
addNums(5)

addNums(5) = 5 + addNums(4)
addNums(4) = 4 + addNums(3)
addNums(3) = 3 + addNums(2)
addNums(2) = 2 + addNums(1)
addNums(1) = 1


15

#### Writing a Factorial Function

In [65]:
# 5! = 5 * 4 * 3 * 2 * 1
def factorialFunc(num):
    if num <= 1: #base case
        return 1
    else:
        return num * factorialFunc(num - 1)
factorialFunc(5)
#check this out with python

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [68]:
#return num - subNums(num - 1)
#subNums(10)
numbers = [4,11,20,3,15,20]

def subNums(num):
    if num <=1:
        return num
    else:
        return num - subNums(num-1)

#         print(f'number: {num} - {numbers}')

subNums(10)

5

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [70]:
def my_range(stop,start,step=2):
    while start < stop:
        yield start #like returning but not... 
                    #yield denoes a generator
        start += step
        
for i in my_range(20, start = 2):
    my_generator_value = i
    print(my_generator_value)

2
4
6
8
10
12
14
16
18


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [80]:
nums = [3,4,5,6]
squares = list(map(lambda num: num ** 2, nums))
print(squares)

[9, 16, 25, 36]


In [82]:
def squaresGen(stop, start, step = 1):
    while start >= stop:
        yield start ** 2
        start -= step
        
for i in squaresGen(0, start = 6):
    print(i)


36
25
16
9
4
1
0


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [96]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

def a_places(place):
    if place.strip() == "":
        return False
    else:
        return True
new_places_list = list(filter(a_places, places))
print(new_places_list)

['Argentina', 'San Diego', 'Boston', 'New York']


In [97]:
new_places_lambda = list(filter(lambda p: False if p.strip() == "" else True, places))
print(new_places_lambda)

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [98]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key = lambda name: name.split(" ")[-1].lower())
print(author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [103]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
#def conversion(places):

#fahr = list(map(lambda x: (9/5)*C +32, places))

fahr = list(map(lambda x: (x[0], (9/5)*x[1]+32), places))
print(fahr)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 

Iteration 0: 1

Iteration 1: 1

Iteration 2: 2

Iteration 3: 3

Iteration 4: 5

Iteration 5: 8`

In [106]:
def fib(num): #base case
    if num == 0:
        return 1
    elif num == 1:
        return 1
    else:
        return fib(num-1) + fib(num-2)

for i in range(6):
    print(f'Iteration {(i)} : {fib(i)}')

Iteration 0 : 1
Iteration 1 : 1
Iteration 2 : 2
Iteration 3 : 3
Iteration 4 : 5
Iteration 5 : 8
